Add diagrams for FLP and rework overview/extensions (#399)
cfrg/draft-irtf-cfrg-vdaf: draft-irtf-cfrg-vdaf.md

proof system and some extensions to it. The construction is specified in
{{flp-bbcggi19-construction}}.

### Overview {#flp-bbcggi19-overview}

Conventional zero-knowledge proof systems involve two parties:

* The prover, who holds a measurement and generates a proof of the
  measurement's validity

* The verifier who holds an encryption of, or commitment to, the measurement
  and checks the proof

Our proof system is much the same, except the verifier is split across multiple
Aggregators, each of which has a secret share of the measurement rather than an
encryption of it.

Validity is defined in terms of an arithmetic circuit evaluated over the
measurement. The inputs to this circuit are elements of a finite field that
comprise the encoded measurement; the gates of the circuit are multiplication,
addition, and subtraction operations; and the output of the circuit is a single
field element. If the value is zero, then the measurement is deemed valid;
otherwise, if the output is non-zero, then the measurement is deemed invalid.

For example, the simplest circuit specified in this document is the following
({{prio3count}}):

~~~
C(x) = x * (x-1)
~~~

> This circuit contains one subtraction gate (`x -1`) and one multiplication
> gate (`x * (x -1)`). Observe that `C(x) = 0` if and only if `x in range(2)`.

Reviewer: nit: formatting
```suggestion
This circuit contains one subtraction gate (`x - 1`) and one multiplication
gate (`x * (x - 1)`). Observe that `C(x) = 0` if and only if `x in range(2)`.
```

Our goal is to allow each Aggregator, who holds a secret share of `x`, to
correctly compute a secret share of `C(x)`. This allows the Aggregators to
determine validity by combining their shares of the output.

Suppose for a moment that the validity circuit `C` is affine, meaning its only
operations are addition, subtraction, and multiplication-by-constant. (The
circuit above is non-affine because it contains a multiplication gate with

> non-constant inputs.) Then each Aggregator can compute its share locally, since

Reviewer: Note that one constant input and one non-constant input is affine
```suggestion
two non-constant inputs.) Then each Aggregator can compute its share locally, since
```

~~~
C(x_shares[0] + ... + x_shares[SHARES-1]) =
    C(x_shares[0]) + ... + C(x_shares[SHARES-1])
~~~

(Note that, for this equality to hold, it is necessary to scale any addition of
a constant in the circuit by `1/SHARES`.) However, this is not the case if `C`

> contains multiplication gates with non-constant inputs. Thus our goal is to

Reviewer: Same as above
```suggestion
contains multiplication gates with two non-constant inputs. Thus our goal is to
```

a constant in the circuit by `1/SHARES`.) However, this is not the case if `C`
contains multiplication gates with non-constant inputs. Thus our goal is to
transform these multiplication gates into computations on secret shared data
that each Aggregator can perform locally.

The key idea is to have the prover construct a polynomial `p` such that `p(j)`
is equal to the output of the `j`-th multiplication gate. Polynomial evaluation
is fully linear, which means the coefficients of the polynomial can be secret
shared in a way that allows each Aggregator to compute a share of `p(j)` for
any `j`. These intermediate results can then be combined with the affine
operations of the validity circuit to produce the final output.

Applying this idea to the example circuit `C` above:

1. The Client, given its measurement `x`, constructs the lowest degree
   polynomial `p` for which `p(0) = s` and `p(1) = x * (x - 1)`, where `s` is a
   random blinding value generated by the Client. (The blinding value is to
   protect the privacy of the measurement.) It sends secret shares of `x` and
   the coefficients of `p` to each of the Aggregators.

> 1. Each Aggregator locally computes and broadcasts its share of `p(0)`, which

Reviewer: This needs to be updated to match the above changes.
```suggestion
1. Each Aggregator locally computes and broadcasts its share of `p(1)`, which
```

   is equal to its share of `C(x)`.

In fact, our FLP is slightly more general than this. We can replace the
multiplication gate with any non-affine sub-circuit and apply the same idea.
For example, in {{prio3sum}}, the validity circuit uses the following
sub-circuit multiple times:

~~~

> Range2(x) = x * (x -1) = x^2 - x

Reviewer: nit: formatting
```suggestion
Range2(x) = x * (x - 1) = x^2 - x
```

## The FLP of {{BBCGGI19}} {#flp-bbcggi19}

This section describes an FLP based on the construction from {{BBCGGI19}},
Section 4.2. We begin in {{flp-bbcggi19-overview}} with an overview of their
proof system and some extensions to it. The construction is specified in
{{flp-bbcggi19-construction}}.

### Overview {#flp-bbcggi19-overview}

Conventional zero-knowledge proof systems involve two parties:

* The prover, who holds a measurement and generates a proof of the
  measurement's validity

* The verifier who holds an encryption of, or commitment to, the measurement
  and checks the proof

Our proof system is much the same, except the verifier is split across multiple
Aggregators, each of which has a secret share of the measurement rather than an
encryption of it.

Validity is defined in terms of an arithmetic circuit evaluated over the
measurement. The inputs to this circuit are elements of a finite field that
comprise the encoded measurement; the gates of the circuit are multiplication,

> addition, and negation operations; and the output of the circuit is a single

Reviewer: This formulation of a circuit seems unusual, I think either 'multiplication and addition' or 'multiplication, addition, and subtraction' would be more conventional.

Author: Hmm, I think of addition, multiplication, and negation as analogues of OR, AND, and NOT respectively for boolean circuits, which to me is more intuitive. What about stating this analogy here, and maybe me mentioning that subtraction is negation followed by addition?

Reviewer: I don't think we need to draw direct parallels to boolean circuits. Besides, if we were to project a boolean circuit into an arithmetic circuit, logical negation would probably be implemented with (1 - x), so I think the analogy doesn't hold up.

Looking at the 2019 Prio paper, it says in section 7.1 that arithmetic circuits over a ring R "may contain addition, subtraction and multiplication gates, as well as constants from R." The Wikipedia page for [Arithmetic circuit complexity](https://en.wikipedia.org/wiki/Arithmetic_circuit_complexity) says that circuits are made up of input gates, sum gates, and product gates.

Author: I'm not sure I aqgree, but I don't feel strongly either way. I replaced "negation" with "subtraction".

> field element. If the value is zero, then the measurement is deemed valid;

Author: This is introduced in the extensions section, in the very first paragraph. Do you think we should either introduce it here or say more about it there? I figured we should keep the overview as conceptually close to the paper as possible.

> field element. If the value is zero, then the measurement is deemed valid;

Reviewer: I missed it on my first read (probably because it wasn't part of the diff), but this is addressed as the fourth extension in 7.3.1.1. LGTM.

otherwise, if the output is non-zero, then the measurement is deemed invalid.

For example, the simplest circuit specified in this document is the following
({{prio3count}}):

~~~
C(x) = x * (x-1)
~~~

This circuit contains one negation gate (`-1`), one addition gate (`x + (-1)`),
and one multiplication gate (`x * (x + (-1))`). Observe that `C(x) = 0` if and
only if `x in range(2)`.

Our goal is to allow each Aggregator, who holds a secret share of `x`, to
correctly compute a secret share of `C(x)`. This allows the Aggregators to
determine validity by combining their shares of the output.

Suppose for a moment that the validity circuit `C` is affine, meaning its only
operations are negation, addition, and multiplication-by-constant. (The circuit
above is non-affine because it contains a multiplication gate with non-constant
inputs.) Then each Aggregator can compute its share locally, since

~~~
C(x_shares[0] + ... + x_shares[SHARES-1]) =
    C(x_shares[0]) + ... + C(x_shares[SHARES-1])
~~~

> (Note that, for this equality to hold, it is necessary to scale any constants

Reviewer: Constants used in constant multiplications are fine as-is, but constants that get added need to be scaled, since this is one way of replacing them with additive shares of themselves.
```suggestion
(Note that, for this equality to hold, it is necessary to scale any addition of a constant
```

Author: Good catch.

> in the circuit by `1/SHARES`.) However, this is not the case if `C` is

Reviewer: nit: typo
```suggestion
in the circuit by `1/SHARES`.) However, this is not the case if `C` has
```

> multiplication gates with non-constant inputs. Thus or goal is to transform

Reviewer: nit: typo
```suggestion
multiplication gates with non-constant inputs. Thus our goal is to transform
```

these multiplication gates into computations on secret shared data that each
Aggregator can perform locally.

The key idea is to have the prover construct a polynomial `p` such that `p(j)`
is equal to the output of the `j`-th multiplication gate. Polynomial evaluation
is fully linear, which means the coefficients of the polynomial can be secret
shared in a way that allows each Aggregator to compute a share of `p(j)` for
any `j`. These intermediate results can then be combined with the affine
operations of the validity circuit to produce the final output.

Applying this idea to the example circuit `C` above:

1. The Client, given its measurement `x`, constructs the lowest degree
   polynomial `p` for which `p(0) = x * (x - 1)`. It sends secret shares of `x`

>    and the coefficients of `p` to each of the Aggregators.

Reviewer: This isn't quite right, because the wire polynomials evaluated at the first point produce blinding values. (and the gadget polynomial evaluated at the first point outputs some function of the blinding values) I think we should probably mention or allude to the blinding here, since it's necessary for zero knowledge, and it's not mentioned below.

Author: Nice catch :) Please take another look: I used `s` to denote the blinding factor, which is short for `seed`, the word we use in the actual specification.
